### ISSUE TITLE
windows buildbot errors,  already imported failures

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2065]
|<img alt="scmikes's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| scmikes reported|
|-|-|
|Trac ID|trac#2065|
|Type|defect|
|Created|2006-09-07 16:31:22Z|

```
ERROR]: twisted.test.test_modules.BasicTests.test_loadPackagesAndModules

Traceback (most recent call last):
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\test\test_modules.py", line 73, in test_loadPackagesAndModules
    self.failUnlessIdentical(
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\test\test_modules.py", line 27, in findByIteration
    for modinfo in where.walkModules(importPackages=importPackages):
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\modules.py", line 602, in walkModules
    for package in self.iterModules():
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\modules.py", line 596, in iterModules
    for module in entry.iterModules():
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\modules.py", line 118, in iterModules
    for potentialTopLevel in placeToLook.children():
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\filepath.py", line 59, in children
    return map(self.child, self.listdir())
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\filepath.py", line 291, in listdir
    return listdir(self.path)
exceptions.WindowsError: [Errno 267] The directory name is invalid: 'c:\\python24\\lib\\site-packages\\sasync-0.1-py2.4.egg/*.*'
===============================================================================
[ERROR]: twisted.test.test_modules.BasicTests.test_onlyTopModules

Traceback (most recent call last):
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\test\test_modules.py", line 55, in test_onlyTopModules
    for module in modules.iterModules():
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\modules.py", line 596, in iterModules
    for module in entry.iterModules():
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\modules.py", line 118, in iterModules
    for potentialTopLevel in placeToLook.children():
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\filepath.py", line 59, in children
    return map(self.child, self.listdir())
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\filepath.py", line 291, in listdir
    return listdir(self.path)
exceptions.WindowsError: [Errno 267] The directory name is invalid: 'c:\\python24\\lib\\site-packages\\sasync-0.1-py2.4.egg/*.*'
===============================================================================
[ERROR]: twisted.test.test_modules.PathModificationTest.test_listingModules

Traceback (most recent call last):
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\test\test_modules.py", line 163, in test_listingModules
    self._listModules()
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\test\test_modules.py", line 152, in _listModules
    pkginfo = modules.getModule(self.packageName)
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\modules.py", line 625, in getModule
    return theSystemPath[moduleName]
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\modules.py", line 584, in __getitem__
    for module in self.iterModules():
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\modules.py", line 596, in iterModules
    for module in entry.iterModules():
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\modules.py", line 118, in iterModules
    for potentialTopLevel in placeToLook.children():
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\filepath.py", line 59, in children
    return map(self.child, self.listdir())
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\filepath.py", line 291, in listdir
    return listdir(self.path)
exceptions.WindowsError: [Errno 267] The directory name is invalid: 'c:\\python24\\lib\\site-packages\\sasync-0.1-py2.4.egg/*.*'
===============================================================================
[ERROR]: twisted.test.test_modules.PathModificationTest.test_underUnderPathAlreadyImported

Traceback (most recent call last):
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\test\test_modules.py", line 138, in test_underUnderPathAlreadyImported
    self._underUnderPathTest()
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\test\test_modules.py", line 127, in _underUnderPathTest
    modinfo = modules.getModule(self.packageName)
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\modules.py", line 625, in getModule
    return theSystemPath[moduleName]
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\modules.py", line 584, in __getitem__
    for module in self.iterModules():
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\modules.py", line 596, in iterModules
    for module in entry.iterModules():
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\modules.py", line 118, in iterModules
    for potentialTopLevel in placeToLook.children():
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\filepath.py", line 59, in children
    return map(self.child, self.listdir())
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\filepath.py", line 291, in listdir
    return listdir(self.path)
exceptions.WindowsError: [Errno 267] The directory name is invalid: 'c:\\python24\\lib\\site-packages\\sasync-0.1-py2.4.egg/*.*'
===============================================================================
[ERROR]: twisted.test.test_modules.ZipPathModificationTest.test_listingModules

Traceback (most recent call last):
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\test\test_modules.py", line 163, in test_listingModules
    self._listModules()
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\test\test_modules.py", line 152, in _listModules
    pkginfo = modules.getModule(self.packageName)
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\modules.py", line 625, in getModule
    return theSystemPath[moduleName]
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\modules.py", line 584, in __getitem__
    for module in self.iterModules():
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\modules.py", line 596, in iterModules
    for module in entry.iterModules():
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\modules.py", line 118, in iterModules
    for potentialTopLevel in placeToLook.children():
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\filepath.py", line 59, in children
    return map(self.child, self.listdir())
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\filepath.py", line 291, in listdir
    return listdir(self.path)
exceptions.WindowsError: [Errno 267] The directory name is invalid: 'c:\\python24\\lib\\site-packages\\sasync-0.1-py2.4.egg/*.*'
===============================================================================
[ERROR]: twisted.test.test_modules.ZipPathModificationTest.test_underUnderPathAlreadyImported

Traceback (most recent call last):
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\test\test_modules.py", line 138, in test_underUnderPathAlreadyImported
    self._underUnderPathTest()
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\test\test_modules.py", line 127, in _underUnderPathTest
    modinfo = modules.getModule(self.packageName)
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\modules.py", line 625, in getModule
    return theSystemPath[moduleName]
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\modules.py", line 584, in __getitem__
    for module in self.iterModules():
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\modules.py", line 596, in iterModules
    for module in entry.iterModules():
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\modules.py", line 118, in iterModules
    for potentialTopLevel in placeToLook.children():
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\filepath.py", line 59, in children
    return map(self.child, self.listdir())
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\python\filepath.py", line 291, in listdir
    return listdir(self.path)
exceptions.WindowsError: [Errno 267] The directory name is invalid: 'c:\\python24\\lib\\site-packages\\sasync-0.1-py2.4.egg/*.*'
===============================================================================
[ERROR]: twisted.test.test_udp.MulticastTestCase.testMulticast

Traceback (most recent call last):
  File "e:\twistedbuildbot\W32-full2.4-scmikes-select\Twisted\twisted\test\test_udp.py", line 679, in testMulticast
    self.assertEquals(self.server.packets[0][0], "hello world")
exceptions.IndexError: list index out of range
-------------------------------------------------------------------------------
```

<details><summary>Searchable metadata</summary>

```
trac-id__2065 2065
type__defect defect
reporter__scmikes scmikes
priority__highest highest
milestone__Core_2_5 Core-2.5
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__error_buildbot_win32 error buildbot win32
time__1157646682000000 1157646682000000
changetime__1163563327000000 1163563327000000
version__None None
owner__ 
cc__glyph cc__radix cc__scmikes cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)<a name="note_3"></a>|@radix commented|
|-|-|

Can someone please explain what is going on here.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph set owner to @teratorn|
|-|-|

I am not the windows maintainer.

**Comment by twisted-trac** [comment]
|<img alt="scmikes's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|scmikes commented|
|-|-|

This defect is triggered by the async egg.  This seems to be a problem with reading from the egg.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to TimothyFitz|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun removed owner|
|-|-|

In modules-listdir-2065, t.p.modules now has tests for the case where sys.path contains a path which does not exist as well as a path which refers to a regular file rather than a directory, and performs error handling around the FilePath.children call to notice and ignore these cases.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_10"></a>|@jml set owner to @exarkun|
|-|-|

Looks good. Glad that the windows stuff got moved from plugin.py to win32.py

Only complaint is that the phrase "does not prevent `walkModules` from producing other results" (iin docstrings for tests) is quite vague and the earlier negative makes it difficult to parse. Perhaps "Check that sys.modules ignores path names in sys.path that refer to ___" instead.

Please fix and then merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set status to `closed`|
|-|-|

(In [18695]) Merge modules-listdir-2065-2

Author: exarkun
Reviewer: jml
Fixes [#2065](https://github.com/twisted/twisted/issues/2065)

Handle non-existent and non-directory entries in a plugin package's
__path__ attribute.  Tests for same.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
